Add current version to the upgrade message of the Flutter tool

karamage opened the pull request:
## Description

Add current version to the upgrade message of the Flutter tool. 

I'm a first-time contributor.
There are many things I don't understand, so please let me know.
Please point out any mistakes.
Thank you for reviewing it.

```
Upgrading Flutter to 1.22.1 from 1.22.0 in C:\Programming\Flutter\sdk\flutter...                                                            Checking Dart SDK version...                                                                                            Downloading Dart SDK from Flutter engine c8e3b9485386425213e2973126d6f57e7ed83c54...                                    Unzipping Dart SDK...
Building flutter tool...
Running pub upgrade...

Upgrading engine...
```

## Related Issues

Add current version to the upgrade message of the Flutter tool. #63023

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behavior with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

jonahwilliams commented:
@karamage can you sign the CLA?

karamage commented:
@jonahwilliams 
OK.
 I signed the CLA.

karamage commented:
@googlebot I signed it!

jonahwilliams commented on packages/flutter_tools/lib/src/commands/upgrade.dart:
```diff
@@ -134,6 +134,7 @@ class UpgradeCommandRunner {
       );
     }
     recordState(flutterVersion);
+    globals.printStatus('Upgrading Flutter to ${upstreamRevision} from ${flutterVersion.frameworkRevision} ($workingDirectory...)');
```
We don't currently have any tests for what gets printed during the upgrade. The closest example is one that checks that a message is printed if nothing is upgraded:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart#L92

I would see if you adapt that to add coverage for this.

karamage replied:
Thanks for reviewing it for you.
I agree with you about adding a test.
I'll TRY to do it.

karamage replied:
I'm sorry.
I tried, but I can't add the test.
I want to check the standard output, please tell me how to do it.

Reference.
 > expect(testLogger.statusText, contains('Flutter is already up to date'));

I ran the test and the testLogger.statusText was empty.

Please let me know.

RoyARG02 replied:
@karamage check if the status text you want to display is actually being displayed, like these lines in a similar test: 

https://github.com/flutter/flutter/blob/fb508db94d9ecfac3b21f961bfd835f77eca9b88/packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart#L115-L117

I think you should take this (similar)test as a reference to create your new test:

https://github.com/flutter/flutter/blob/fb508db94d9ecfac3b21f961bfd835f77eca9b88/packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart#L310-L314

karamage replied:
@RoyARG02 Thank you. I'll try.

karamage replied:
@RoyARG02 I add the test to test/commands.shard/permeable/upgrade_test.dart L310.
But the test fails.

```dart
testUsingContext('Show current version to the upgrade message.', () async {
        final Future<FlutterCommandResult> result = fakeCommandRunner.runCommand(
          force: false,
          continueFlow: false,
          testFlow: false,
          gitTagVersion: gitTagVersion,
          flutterVersion: flutterVersion,
          verifyOnly: false,
        );
        expect(await result, FlutterCommandResult.success());
        expect(testLogger.statusText, contains('Upgrading Flutter to'));
        expect(testLogger.statusText, contains('from ${flutterVersion.frameworkRevision} in'));
        expect(processManager.hasRemainingExpectations, isFalse);
      }, overrides: <Type, Generator>{
        ProcessManager: () => processManager,
        Platform: () => fakePlatform,
      });
```

```
00:05 +12: UpgradeCommandRunner runs upgrade Show current version to the upgrade message.
Expected: contains 'Upgrading Flutter to'
  Actual: ''
```

testLogger.statusText was empty.
Please tell me where I'm wrong.
Thanks for the help.

jonahwilliams replied:
Do you know how to run the test locally? There are some instructions in `packages/flutter_tools/README.md`. You should be able to run this with `dart test test/commands.shard/permeable/upgrade_test.dart`.

Once you have that running, start with something simple like adding `print('here');` to the tool code to make sure it is running what you think it is running

karamage replied:
I read the README and ran the test locally.

... /... /bin/dart pub run test test test/commands.shard/permeable/upgrade_test.dart

But the test failed.


I will learn how to test using the simple method you have taught me.
Thank you.

jonahwilliams replied:
Try this out for a bit and let me know if you make any progress. I will take a look at this PR again tomorrow and make some suggestions to get your test working correctly

RoyARG02 replied:
@karamage I think you need to use `realCommandRunner` in this case, as `fakeCommandRunner.upgradeChannel` doesn't do anything:
https://github.com/flutter/flutter/blob/fb508db94d9ecfac3b21f961bfd835f77eca9b88/packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart#L401

It also explains you getting **`Actual: ''`** in your test.

karamage replied:
Thank you. I appreciate it.

Okay, I understand.

I will try to write the test code using realCommandRunner.

karamage replied:
@RoyARG02 @jonahwilliams 

I've added the test code.

I would like you to review it.

Thank you for showing me how to test it.

The test was a success.

```
../../bin/dart pub run test test/commands.shard/permeable/upgrade_test.dart
00:04 +15: All tests passed!
```

jonahwilliams replied:
🎊

jonahwilliams commented on packages/flutter_tools/lib/src/commands/upgrade.dart:
```diff
@@ -243,7 +244,6 @@ class UpgradeCommandRunner {
   /// If the user is on a deprecated channel, attempts to migrate them off of
   /// it.
   Future<void> upgradeChannel(FlutterVersion flutterVersion) async {
-    globals.printStatus('Upgrading Flutter from $workingDirectory...');
```
I would leave this message in upgradeChannel, and instead pass the upstreamRevision revision in as an argument.

karamage replied:
Understood. I will modify it to pass in arguments.

jonahwilliams commented on packages/flutter_tools/lib/src/commands/upgrade.dart:
```diff
@@ -134,6 +134,7 @@ class UpgradeCommandRunner {
       );
     }
     recordState(flutterVersion);
+    globals.printStatus('Upgrading Flutter to ${upstreamRevision} from ${flutterVersion.frameworkRevision} ($workingDirectory...)');
```
For just a single identifier you don't need the`{}`:

```suggestion
    globals.printStatus('Upgrading Flutter to $upstreamRevision from ${flutterVersion.frameworkRevision} ($workingDirectory...)');
```

karamage replied:
I'll fix it.

jonahwilliams reviewed (commented):
Seems like a nice change! If you can add a test case I'd be happy to land this

RoyARG02 commented on packages/flutter_tools/lib/src/commands/upgrade.dart:
```diff
@@ -242,8 +242,8 @@ class UpgradeCommandRunner {
-  Future<void> upgradeChannel(FlutterVersion flutterVersion) async {
-    globals.printStatus('Upgrading Flutter from $workingDirectory...');
+  Future<void> upgradeChannel(FlutterVersion flutterVersion, String upstreamRevision) async {
+    globals.printStatus('Upgrading Flutter to $upstreamRevision from ${flutterVersion.frameworkRevision} ($workingDirectory...)');
```
```suggestion
    globals.printStatus('Upgrading Flutter to $upstreamRevision from ${flutterVersion.frameworkRevision} in $workingDirectory...');
```

karamage commented:
@googlebot I consent.

RoyARG02 commented:
@googlebot I consent.

karamage committed:
add test for upgradeChannel. #63023

jonahwilliams reviewed (dismissed):
LGTM!

I'd also like @christopherfujino to double check too. Thank you @RoyARG02 for the help as well

RoyARG02 commented on packages/flutter_tools/lib/src/commands/upgrade.dart:
```diff
@@ -262,8 +262,8 @@ class UpgradeCommandRunner {
-  Future<void> upgradeChannel(FlutterVersion flutterVersion) async {
-    globals.printStatus('Upgrading Flutter from $workingDirectory...');
+  Future<void> upgradeChannel(FlutterVersion flutterVersion, String upstreamRevision) async {
+    globals.printStatus('Upgrading Flutter to $upstreamRevision from ${flutterVersion.frameworkRevision} in $workingDirectory...');
```
I just realized that using `flutterVersion.frameworkRevision` is probably not what the PR is trying to do:

Take a look at these lines of code:
https://github.com/flutter/flutter/blob/13896b3bd1858687a06b1639d60c877206170df8/packages/flutter_tools/lib/src/version.dart#L170-L183

My understanding is that `flutterVersion.frameworkVersion` should be used, but is there anyway the corresponding version of the `upstreamRevision` can be obtained?

jonahwilliams replied:
Thats a good point, it is probably printing out the commit hash right now. We should know what the tag we're checking out is

RoyARG02 replied:
This might be something of interest in that case:
https://github.com/flutter/flutter/blob/13896b3bd1858687a06b1639d60c877206170df8/packages/flutter_tools/lib/src/version.dart#L77

RoyARG02 replied:
> Thats a good point, it is probably printing out the commit hash right now. We should know what the tag we're checking out is

@jonahwilliams You're right. #68866 was merged yesterday which uses `flutterVersion.frameworkRevision`:
```
$ flutter upgrade --verify-only
A new version of Flutter is available on channel master

The latest revision: 18341efa2fb8c3570ca32e71d1a9da74e8b2cbf5
Your current version: 8b9e9680d24ffd4b3dbfa216f5f64a0490088f45

To upgrade now, run "flutter upgrade".
```

karamage replied:
@RoyARG02 @jonahwilliams Thanks for reviewing it for you.
I understand that using flutterVersion.frameworkRevision is a mistake.
Thank you.

I will modify it to use frameworkVersion.

karamage replied:
I have modified it to use frameworkVersion.
I also rewrote the test and it was successful!
Please give you a review.

christopherfujino replied:
I'm taking a look now.

karamage committed:
Fix to use frameworkVersion #63023

christopherfujino commented on packages/flutter_tools/lib/src/commands/upgrade.dart:
```diff
@@ -154,7 +154,7 @@ class UpgradeCommandRunner {
     }
     recordState(flutterVersion);
-    await upgradeChannel(flutterVersion);
+    await upgradeChannel(flutterVersion, gitTagVersion.frameworkVersionFor(upstreamRevision));
```
Unfortunately, I don't believe this is right either. The `.frameworkVersionFor()` method is confusingly named. It does not return the correct version for the input revision you input, but rather the `GitTagVersion` that it is a method for. And on line 157, I believe that `gitTagVersion` is the current HEAD, or the version that the tool is currently at.

Please see my comments at @RoyARG02's PR for more context: https://github.com/flutter/flutter/pull/69420/files#r515275570. It seems like are both trying to hook into functionality that we don't yet support.

RoyARG02 replied:
@karamage I've updated #69420 as per suggestions from @christopherfujino. Please have a look at it(I've not updated the tests yet).

If you need any clarifications about my approach, or have a better idea on implementing the feature, we can discuss on the chat.

karamage replied:
@RoyARG02 @christopherfujino Thanks for your suggestion.
I'll look at #69420.

I am trying to understand your approach.

RoyARG02 commented:
@karamage #69420 has landed. You can work on this PR now.

karamage commented:
@RoyARG02 Thank you.  I will work on this PR.

karamage committed:
tweak.

karamage commented:
I went back to the changes because of the conflicts.

I will continue to change the PR from here.

It's going to take some time to understand the changes.

karamage committed:
Show current version to the upgrade message.

karamage committed:
Remove unnecessary parts in test code

karamage commented:
I saw #69420. It was very informative. Thank you.

I wrote the code for this PR, but I'm not sure if it's good enough.

The upgradeChannel method is gone, so I made a change to the recordState method.

If I'm wrong, please let me know.

I wrote the test code and ran the test.
The test was successful.

Thank you for reading.

RoyARG02 commented on packages/flutter_tools/lib/src/commands/upgrade.dart:
```diff
@@ -151,14 +151,15 @@ class UpgradeCommandRunner {
 
-  void recordState(FlutterVersion flutterVersion) {
+  void recordState(FlutterVersion flutterVersion, FlutterVersion upstreamVersion) {
+    globals.printStatus('Upgrading Flutter to ${upstreamVersion.frameworkVersion} from ${flutterVersion.frameworkVersion} in $workingDirectory...');
```
It might be better if you directly put this line in `runCommandFirstHalf` after calling `recordState`.

RoyARG02 commented on packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart:
```diff
@@ -282,6 +282,43 @@ void main() {
+      const String upstreamVersion = '4.5.6';
+
+      when(flutterVersion.frameworkRevision).thenReturn(revision);
+      when(flutterVersion.frameworkRevisionShort).thenReturn(revision);
```
I don't think you'll need `flutterVersion.frameworkRevisionShort` in your test.
Same for `latestVersion.frameworkRevisionShort` below.

RoyARG02 commented on packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart:
```diff
@@ -282,6 +282,43 @@ void main() {
+        ),
+      );
+      realCommandRunner.recordState(flutterVersion, latestVersion);
+      expect(testLogger.statusText, contains('Upgrading Flutter to ${latestVersion.frameworkVersion} from ${flutterVersion.frameworkVersion} in ${realCommandRunner.workingDirectory}'));
```
Tests should be matched against hardcoded strings.
See https://github.com/flutter/flutter/pull/69420#discussion_r519069628 for context.

karamage committed:
fix the points pointed out in the review

karamage commented:
@RoyARG02 Your suggestions were very helpful. Thank you.
I have modified the code according to your suggestion.
I would appreciate it if you could take a look at the code again.

RoyARG02 commented on packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart:
```diff
@@ -282,6 +282,51 @@ void main() {
+        ]),
+      ]);
+
+      await realCommandRunner.runCommandFirstHalf(force: true, gitTagVersion: gitTagVersion, flutterVersion: flutterVersion, testFlow: true, verifyOnly: false);
```
Since the message you are expecting is in `runCommandFirstHalf`, it might be preferrable to run `fakeCommandRunner.runCommand` and awaiting for `FlutterCommandResult.success` instead of adding commands explicitly to `processManager`, just like other tests.

Also, please format the call as following: 
```dart
final Future<FlutterCommandResult> result = fakeCommandRunner.runCommand(
        force: false,
        continueFlow: false,
        testFlow: false,
        gitTagVersion: gitTagVersion,
        flutterVersion: flutterVersion,
        verifyOnly: true,
      );
```

RoyARG02 commented on packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart:
```diff
@@ -282,6 +282,51 @@ void main() {
+      ]);
+
+      await realCommandRunner.runCommandFirstHalf(force: true, gitTagVersion: gitTagVersion, flutterVersion: flutterVersion, testFlow: true, verifyOnly: false);
+      expect(testLogger.statusText, contains('Upgrading Flutter to 4.5.6 from 1.2.3 in null'));
```
Could you change `null` to something meaningful? I think using `fakeCommandRunner.workingDirectory` might fix this.

karamage committed:
improve test code

karamage commented:
@RoyARG02 The improvements you have pointed out have been very helpful.
Thanks to that, I now understand how to use the fakeCommandRunner.

I have fixed the test code, could you take a look at it?
Thank you, I've fixed it and would appreciate it if you could take a look.

RoyARG02 commented:
Everything looks good, let's see how the maintainers respond.

christopherfujino reviewed (approved):
LGTM! Thanks so much @karamage for making this change and @RoyARG02 for helping and doing code review.
